daceml/autodiff/backward_pass_generator.py

```diff
@@ -0,0 +1,1270 @@
+    """Returns true if state has any inplace operations
+    Note that this method is currently much stronger than required; some of the constrains can be
+    loosened in the future.
+    """
+
+    sdfg = state.parent
+
+    # check that each data descriptor has at most one access nodes
+    seen_accesses: typing.Set[str] = set()
+    for node in state.nodes():
+        if isinstance(node, nd.AccessNode):
+            if node.data in seen_accesses:
+                return True
+            seen_accesses.add(node.data)
+
+    # edges with scalar memlets can be used to connect two code nodes together. If this feature is
+    # used, it should be done using a new scalar everytime.
+    # When a scalar is used in a code -> code edge, it should also have an AccessNode that refers to it.
+    seen_scalars = set()
+    for edge in state.edges():
+        memlet_data = edge.data.data
+        if (isinstance(sdfg.arrays[memlet_data], dt.Scalar)
+                and isinstance(edge.src, nd.CodeNode)
+                and isinstance(edge.dst, nd.CodeNode)):
+            if memlet_data in seen_scalars or memlet_data in seen_accesses:
+                return True
+            seen_scalars.add(memlet_data)
+    return False
+
+
+def _get_matching_entry(state: SDFGState, map_exit: nd.MapExit) -> nd.MapEntry:
+    """Get the matching `MapEntry` for a `MapExit`"""
+    cands = [
+        node for node in state.nodes()
+        if isinstance(node, nd.MapEntry) and node.map is map_exit.map
+    ]
+
+    if len(cands) != 1:
+        raise AutoDiffException(
+            "More than one map entry found for map {}".format(map_exit.map))
+    return cands[0]
+
+
+def _walk_up_memlet_tree_through_view_nodes(
+    sdfg, forward_state, start_name
+) -> typing.Tuple[typing.Union[dt.Scalar, dt.Array], str,
+                  typing.Deque[typing.Tuple[str, dt.Data, Memlet]]]:
+    """ Starting from the (singular) access node for ``start_name`` in ``forward_state``, walk up the
+        memlet path until a non-view node is reached
+
+        :param sdfg: the forward sdfg
+        :param forward_state: the forward state
+        :param start_name: the name of the array to start at
+        :return: the descriptor at the root of the path, the name at the root of the path, the list of
+                 array names, view data descriptor and memlets encountered along the path.
+    """
+    forwarded_name = start_name
+    view_nodes_to_clone: typing.Deque[typing.Tuple[
+        str, dt.Data, Memlet]] = collections.deque()
+    if type(sdfg.arrays[start_name]) is dt.View:
+        # this is complicated slightly by views: we need to walk up the memlet path until we reach a
+        # non-view access node. We then need to replicate the sequence of views in the backward SDFG.
+        query = [
+            n for n, _ in forward_state.all_nodes_recursive()
+            if isinstance(n, nd.AccessNode) and n.data == start_name
```

Comment:
are we sure we want to do this recursively?

Reply:
nope, good catch